libc: minimal: errno alignment with newlib

carlescufi opened the pull request:
See commit messages.

Note: This is **not** an API break, since only the numbers have changed but not the name of the macros.

carlescufi committed:
libc: minimal: Align errno.h values with newlib

In order to be able to document the error codes with Doxygen and
Breathe, start by aligning the macro values to those provided by newlib
in include/sys/errno.h:

https://github.com/zephyrproject-rtos/newlib-cygwin/blob/zephyr-newlib-4.1.0/newlib/libc/include/sys/errno.h

Also get rid of the (unused and untrue) ERRMAX definition.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
libc: minimal: Add doxygen formatting to errno.h

Format the file so it can be processed by Doxygen.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
ci: Add an errno.h check to CI

Add a very simple check that verifies that new error numbers are added
according to the errno.h file in newlib, in order to maintain
compatibility with it.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: reference: Add a section for the C standard library

In order to document the error codes, introduce a new reference section
describing the C standard library options available in Zephyr, along
with the definitions of the error codes.

At the same time move (most of) the content from the existing user guide
to the new API reference section, removing the need for a separate user
guide for it.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

gmarull commented on .github/workflows/errno.yml:
```diff
@@ -0,0 +1,32 @@
+
+      - name: Run errno.py
+        run: |
+          export ZEPHYR_BASE=${PWD}
+          ./scripts/ci/errno.py
```
script should take required paths from input, e.g. `--zephyr-path`, `--zephyr-sdk-path`

carlescufi replied:
See my comment below, I am not convinced in this particular case.

gmarull commented on .github/workflows/errno.yml:
```diff
@@ -0,0 +1,32 @@
+
+      - name: checkout
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
+          fetch-depth: 0
```
is this needed?

carlescufi replied:
No, I can remove that.

gmarull commented on .github/workflows/errno.yml:
```diff
@@ -0,0 +1,32 @@
+    runs-on: ubuntu-latest
+    container:
+      image: zephyrprojectrtos/ci:v0.17.1
+      options: '--entrypoint /bin/bash'
```
isn't zephyr CI entry point already bash?

carlescufi replied:
I don't think so. Only the [user image](https://github.com/zephyrproject-rtos/docker-image/blob/master/Dockerfile.user#L20) seems to set an entry point

gmarull replied:
example-application doesn't need that, and uses the same container: https://github.com/teslabs/example-application/blob/main/.github/workflows/build.yml

gmarull commented on scripts/ci/errno.py:
```diff
@@ -0,0 +1,47 @@
+        errnos = []
+        for line in f:
+            m = r.match(line)
+            #print(f'{line}: {m}')
```
oops

gmarull commented on scripts/ci/errno.py:
```diff
@@ -0,0 +1,47 @@
+    minimal = Path("lib/libc/minimal/include/errno.h")
+    newlib = Path("arm-zephyr-eabi/arm-zephyr-eabi/include/sys/errno.h")
+
+    try:
+        minimal = os.environ.get('ZEPHYR_BASE') / minimal
+        newlib = os.environ.get('ZEPHYR_SDK_INSTALL_DIR') / newlib
+    except:
+        print('Environment variable missing')
+        sys.exit(1)
```
Better use script arguments to pass required information

carlescufi replied:
I am not convinced about this. This is common practice in our GitHub actions, especially for core env vars like those, which are expected to be present. Also makes it much simpler to run locally.

gmarull replied:
As a general practice I think it's bad to exclusively depend on env vars to operate. What I usually do is to use them as defaults if set, something like e.g.
```
    parser.add_argument(
        "--foo",
        default=os.environ.get("FOO")
    )
```

gmarull commented on scripts/ci/errno.py:
```diff
@@ -0,0 +1,47 @@
+            sys.exit(1)
+
+    print('errno.h validated correctly')
+    sys.exit(0)
```
redundant

gmarull commented on scripts/ci/errno.py:
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python3
```
some docstrings explaining what this script does will help

gmarull commented on .github/workflows/errno.yml:
```diff
@@ -0,0 +1,32 @@
+          ref: ${{ github.event.pull_request.head.sha }}
+          fetch-depth: 0
+
+      - name: west setup
+        run: |
+          west init -l . || true
+          west update
```
since we are just checking repo files, west init/update is not needed

carlescufi replied:
While you are completely correct, I'd rather leave this in for consistency with other actions. This action only ever runs if `errno.h` is modified, so the overhead is very small.

gmarull replied:
this means we're downloading ~2GB of dependencies for nothing every time

gmarull replied:
we can save > 2min, so I think it definitely needs to be removed

![image](https://user-images.githubusercontent.com/25011557/118275558-dda47900-b4c6-11eb-9bab-e00fc9d3cea2.png)

nashif replied:
agree, we do not need to pull the work just to be checking local files

gmarull reviewed (commented):
A few comments related to CI and checker script

gmarull commented on .github/workflows/errno.yml:
```diff
@@ -0,0 +1,31 @@
+
+      - name: checkout
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
```
can also be removed

gmarull commented on scripts/ci/errno.py:
```diff
@@ -0,0 +1,56 @@
+    for e in minimal:
+        if e[0] not in [x[0] for x in newlib] or e[1] != next(
+            filter(lambda _e: _e[0] == e[0], newlib))[1]:
+            print('Invalid entry in errno.h:')
+            print(f'{e[0]} (with value {e[1]})')
```
nit: `print('...', file=sys.stderr)`

gmarull commented on scripts/ci/errno.py:
```diff
@@ -0,0 +1,56 @@
+        for line in f:
+            m = r.match(line)
+            if m:
+                errnos.append((m.group(1), m.group(2)))
```
nit: `errnos.append(m.groups())`

carlescufi replied:
Not sure if there are more groups defined by a regex match, so I'd rather be explicit here.

gmarull replied:
only 2 are defined in the match expression, so we can't have more

carlescufi replied:
Actually, this doesn't work because the first element in `m.group()` is [the entire match](https://docs.python.org/3/library/re.html#match-objects)

gmarull replied:
Hmm... I get this:
```
In [7]: import re

In [8]: m = re.match(r'^\s*#define\s+([A-Z]+)\s+([0-9]+)', '#define X 0')

In [9]: m.groups()
Out[9]: ('X', '0')
```

carlescufi replied:
Doh, I was using `m.group()` and not `m.groups()`. Fixed now @gmarull

gmarull commented on scripts/ci/errno.py:
```diff
@@ -0,0 +1,56 @@
+        minimal = os.environ.get('ZEPHYR_BASE') / minimal
+        newlib = os.environ.get('ZEPHYR_SDK_INSTALL_DIR') / newlib
+    except Exception as e:
+        print('Environment variable missing')
```
`print(..., file=sys.stderr)`

gmarull commented on scripts/ci/errno.py:
```diff
@@ -0,0 +1,56 @@
+    for e in minimal:
+        if e[0] not in [x[0] for x in newlib] or e[1] != next(
+            filter(lambda _e: _e[0] == e[0], newlib))[1]:
+            print('Invalid entry in errno.h:')
```
`print(..., file=sys.stderr)`

carlescufi commented on doc/reference/libc/index.rst:
```diff
@@ -0,0 +1,54 @@
+The `C standard library`_ is an integral part of any C program, and Zephyr
+provides two implementations for the application to choose from.
+
+The first one, named "minimal libc" is part of the Zephyr code base and provides
```
Actually, I wonder if it'd be better just to get rid of the user guide and move all the info to the new reference section I've added.

stephanosio replied:
Sure, IMO, that would be less confusing.

stephanosio replied:
On second thought, keeping a "Standard C Library" doc under the user guide wouldn't necessarily be a bad thing since it probably is of interest to many people who've just started using Zephyr. Maybe we can keep the briefs in the existing doc and move all the details into this new file?

carlescufi replied:
Thanks for the input @stephanosio. I discussed this with @nashif and we agree that we need to start moving information from individual guides to the main reference section, so I will implement your first suggestion and add a redirect.

nashif replied:
is this going to be part of this PR?

carlescufi replied:
Yes, just added it now @nashif

carlescufi committed:
libc: minimal: Add EOVERFLOW

Add the EOVERFLOW error number, which is already in use by some of
Zephyr's modules.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@gmarull addressed your input on the workflow. Thanks for the review!

gmarull commented on .github/workflows/errno.yml:
```diff
@@ -0,0 +1,24 @@
+    steps:
+      - name: Update PATH for west
+        run: |
+          echo "$HOME/.local/bin" >> $GITHUB_PATH
```
Can also be deleted now

gmarull reviewed (commented):
One last change on the ci workflow

gmarull commented on scripts/ci/errno.py:
```diff
@@ -0,0 +1,56 @@
+    minimal = Path("lib/libc/minimal/include/errno.h")
+    newlib = Path("arm-zephyr-eabi/arm-zephyr-eabi/include/sys/errno.h")
+
+    try:
+        minimal = os.environ.get('ZEPHYR_BASE') / minimal
+        newlib = os.environ.get('ZEPHYR_SDK_INSTALL_DIR') / newlib
+    except Exception as e:
+        print('Environment variable missing', file=sys.stderr)
+        sys.exit(1)
```
```suggestion
    try:
        minimal = os.environ['ZEPHYR_BASE'] / minimal
        newlib = os.environ['ZEPHYR_SDK_INSTALL_DIR'] / newlib
    except KeyError as e:
        print(f'Environment variable missing: {e}', file=sys.stderr)
        sys.exit(1)
```

gmarull reviewed (approved):
LGTM!

stephanosio reviewed (approved):
C library related changes LGTM.
